src/mumble/PulseAudio.cpp

```diff
@@ -371,6 +377,18 @@ void PulseAudioSystem::eventCallback(pa_mainloop_api *api, pa_defer_event *) {
 	}
 }
 
+void PulseAudioSystem::corkStream(const bool corked)
+{
+	// TODO: Do I need this lock?
+	if (pasInput)
+	{
+		pa_threaded_mainloop_lock(pam);
+		pa_stream_cork(pasInput, corked, NULL, NULL);
+		pa_threaded_mainloop_unlock(pam);
+	}
```

Comment:
This is something @davidebeatrici might be able to answer.

Reply:
Also is `pa_stream_cork` guaranteed to never throw an exception? Aka is it guaranteed that the release of the lock will be executed in all situations?

Reply:
[`pa_stream_cork`](https://freedesktop.org/software/pulseaudio/doxygen/stream_8h.html#a14e698233ac2d246646651955ab0ec7b) is a C function, so it should never throw.

Reply:
Ah okay. That's fine then :+1:

Reply:
Reading https://freedesktop.org/software/pulseaudio/doxygen/threaded_mainloop.html, I strongly suspect this look indeed has to be aquired

Reply:
I'm pretty sure there's no need, because PulseAudio probably discards anything that goes inside the stream while it's corked.

Reply:
@davidebeatrici, I have seen this lock acquired elsewhere during corking, in places like [chrome OS](https://chromium.googlesource.com/chromiumos/platform/speech_synthesis/+/3728f3b94272b44780deb1a2eee9fc97b4f1170b/linux/audio_output_pulseaudio.cc#184).

The documentation referenced by @Krzmbrzl indicates that

> … a locking scheme must be used to guarantee safe usage … the lock needs to be held whenever you call any PulseAudio function that uses an object associated with this main loop

If `pa_stream *pasInput` is used within the threaded main loop, we need to acquire the lock before we access it. The cork/uncork operation might not be atomic with respect to anything else that is happening, and there are no guarantees about anything else it might touch—like buffer pointers.

I think we're better safe than sorry here.

Reply:
Yes, I agree.

We can always investigate later whether the lock is actually required.